File: Sources/ClaudeRelayClient/PairingController.swift
import Foundation
import ClaudeRelayKit

/// The subset of `RelayConnection` a `PairingController` needs. Lets tests
/// inject a stub without a live socket. `RelayConnection` satisfies it as-is.
@MainActor
public protocol PairingConnection: AnyObject {
    func connect(config: ConnectionConfig, token: String) async throws
    @discardableResult
    func addServerMessageSubscriber(_ handler: @escaping (ServerMessage) -> Void) -> UUID
    func removeSubscriber(_ id: UUID)
    func send(_ message: ClientMessage) async throws
    func disconnect()
}

extension RelayConnection: PairingConnection {}

public enum PairingError: Error, Equatable {
    case invalidCode        // 401
    case rateLimited        // 429
    case tlsRequired        // ws:// to a host ATS blocks
    case unreachable        // transport failure dialing the host
    case timedOut           // no reply within the deadline
    case server(code: Int, message: String)   // any other server error
}

/// Redeems a pairing code over a fresh, pre-auth WebSocket connection and
/// persists the minted token the same way `AddEditServerViewModel.save()` does.
///
/// One home for the redeem sequence so iOS and macOS share it. It deliberately
/// does NOT authenticate — after `pair()` returns, the caller connects through
/// the normal path with the persisted token, keeping one authenticated path.
@MainActor
public final class PairingController {

    private let store: SavedConnectionStore
    private let deviceName: String
    private let platform: String
    private let connectionFactory: () -> PairingConnection
    private let timeout: Duration

    public init(
        store: SavedConnectionStore,
        deviceName: String,
        platform: String,
        connectionFactory: @escaping @MainActor () -> PairingConnection = { RelayConnection() },
        timeout: Duration = .seconds(10)
    ) {
        self.store = store
        self.deviceName = deviceName
        self.platform = platform
        self.connectionFactory = connectionFactory
        self.timeout = timeout
    }

    /// Dials the host in `url`, redeems `url.code`, persists a `ConnectionConfig`
    /// plus the minted token, and returns the config. Throws `PairingError` on any
    /// failure. Never leaves a dangling socket.
    public func pair(_ url: PairingURL) async throws -> ConnectionConfig {
        let connection = connectionFactory()

        // A pre-auth dial: connect() stores the token but does not send
        // auth_request, so an empty token is correct here — we redeem first.
        let dialConfig = ConnectionConfig(
            name: url.host, host: url.host, port: url.port, useTLS: url.useTLS)
        do {
            try await connection.connect(config: dialConfig, token: "")
        } catch {
            throw PairingError.unreachable
        }
        defer { connection.disconnect() }

        let reply = try await redeem(url.code, on: connection)
        guard case .pairSuccess(let token, _, let label) = reply else {
            throw mapError(reply)
        }

        let config = ConnectionConfig(
            name: label.isEmpty ? url.host : label,
            host: url.host, port: url.port, useTLS: url.useTLS)
        store.add(config)
        try? AuthManager.shared.saveToken(token, for: config.id)
        return config
    }

    /// Installs the response subscriber BEFORE sending (mirrors
    /// SessionController.awaitResponse), matches pair_success or error, and
    /// resolves within the deadline. Only one RPC is ever in flight here.
    private func redeem(_ code: String, on connection: PairingConnection) async throws -> ServerMessage {
        let matchTypes: Set<String> = ["pair_success", "error"]

        return try await withThrowingTaskGroup(of: ServerMessage.self) { group in
            let box = ContinuationBox()
            let subId = connection.addServerMessageSubscriber { message in
                guard matchTypes.contains(message.typeString) else { return }
                box.resume(with: message)
            }
            defer { connection.removeSubscriber(subId) }
            defer { group.cancelAll() }

            group.addTask { @MainActor in
                try await withTaskCancellationHandler {
                    try await withCheckedThrowingContinuation { cont in
                        box.attach(cont)
                    }
                } onCancel: {
                    box.cancel()
                }
            }
            group.addTask { @MainActor [timeout] in
                try await Task.sleep(for: timeout)
                throw PairingError.timedOut
            }

            do {
                try await connection.send(.pairRequest(
                    code: code, deviceName: deviceName, platform: platform))
            } catch {
                throw PairingError.unreachable
            }

            guard let result = try await group.next() else { throw PairingError.timedOut }
            return result
        }
    }

    private func mapError(_ message: ServerMessage) -> PairingError {
        guard case .error(let code, let text) = message else {
            return .server(code: -1, message: "Unexpected reply")
        }
        switch code {
        case 401: return .invalidCode
        case 429: return .rateLimited
        default:  return .server(code: code, message: text)
        }
    }
}

/// Bridges the subscriber callback (sync) to the awaiting continuation. The
/// subscriber may fire before or after the continuation is attached, so it
/// stores the first value and replays it on attach. Cancellation-aware: when
/// the task is cancelled, resumes the continuation with CancellationError.
@MainActor
private final class ContinuationBox {
    private var continuation: CheckedContinuation<ServerMessage, Error>?
    private var pending: ServerMessage?
    private var done = false

    func attach(_ cont: CheckedContinuation<ServerMessage, Error>) {
        guard !done else { return }
        if let pending {
            done = true
            cont.resume(returning: pending)
        } else {
            continuation = cont
        }
    }

    func resume(with message: ServerMessage) {
        guard !done else { return }
        if let continuation {
            done = true
            continuation.resume(returning: message)
        } else {
            pending = message
        }
    }

    nonisolated func cancel() {
        Task { @MainActor in
            guard !done else { return }
            done = true
            continuation?.resume(throwing: CancellationError())
        }
    }
}

File: ClaudeRelayAndroid/core-session/src/main/kotlin/relay/session/PairingController.kt
package relay.session

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull
import relay.protocol.ClientMessage
import relay.protocol.ConnectionConfig
import relay.protocol.PairingConnection
import relay.protocol.PairingURL
import relay.protocol.ServerMessage
import java.util.UUID

sealed class PairingError : Exception() {
    object InvalidCode : PairingError()
    object RateLimited : PairingError()
    object TlsRequired : PairingError()
    object Unreachable : PairingError()
    object TimedOut : PairingError()
    data class Server(val code: Int, val text: String) : PairingError()
}

/**
 * Minimal storage seams for injection testing. PairingController uses these
 * interfaces; production passes real stores, tests pass fakes.
 */
interface ConnectionStoreAdapter {
    suspend fun add(connection: ConnectionConfig): List<ConnectionConfig>
}

interface TokenStoreAdapter {
    fun saveToken(token: String, connectionId: UUID)
}

/**
 * Redeems a pairing code over a fresh, pre-auth WebSocket connection and
 * persists the minted token the same way ServersViewModel.addOrUpdate does.
 *
 * Ports the Swift PairingController (Sources/ClaudeRelayClient/PairingController.swift).
 * One home for the redeem sequence so iOS, macOS, and Android share it. It
 * deliberately does NOT authenticate — after pair() returns, the caller connects
 * through the normal path with the persisted token, keeping one authenticated path.
 */
class PairingController(
    private val store: ConnectionStoreAdapter,
    private val tokenStore: TokenStoreAdapter,
    private val deviceName: String,
    private val platform: String,
    private val connectionFactory: () -> PairingConnection,
    private val timeoutMs: Long = 10_000L,
) {
    /**
     * Dials the host in [url], redeems [url.code], persists a ConnectionConfig
     * plus the minted token, and returns the config. Throws [PairingError] on any
     * failure. Never leaves a dangling socket.
     */
    suspend fun pair(url: PairingURL): ConnectionConfig {
        val connection = connectionFactory()

        // Pre-auth dial: connect() with an empty token (we redeem first)
        val dialConfig = ConnectionConfig(
            name = url.host,
            host = url.host,
            port = url.port.toUShort(),
            useTLS = url.useTLS
        )
        try {
            connection.connect(dialConfig, token = "")
        } catch (e: Exception) {
            throw PairingError.Unreachable
        }

        try {
            val reply = redeem(url.code, connection)
            val success = reply as? ServerMessage.PairSuccess ?: throw mapError(reply)
            val config = ConnectionConfig(
                name = success.label.ifEmpty { url.host },
                host = url.host,
                port = url.port.toUShort(),
                useTLS = url.useTLS
            )
            store.add(config)
            tokenStore.saveToken(success.token, config.id)
            return config
        } finally {
            withContext(NonCancellable) {
                connection.disconnect()
            }
        }
    }

    /**
     * Installs the response subscriber BEFORE sending (mirrors
     * SessionController.awaitResponse), matches pair_success or error, and
     * resolves within the deadline. Only one RPC is ever in flight here.
     */
    private suspend fun redeem(code: String, connection: PairingConnection): ServerMessage {
        val deferred = CompletableDeferred<ServerMessage>()
        val matchTypes = setOf("pair_success", "error")

        val subId = connection.addServerMessageSubscriber { msg ->
            if (msg.typeString in matchTypes) {
                deferred.complete(msg)
            }
        }

        try {
            connection.send(ClientMessage.PairRequest(code, deviceName, platform))
            return withTimeoutOrNull(timeoutMs) { deferred.await() } ?: throw PairingError.TimedOut
        } catch (e: CancellationException) {
            throw e
        } catch (e: PairingError) {
            throw e
        } catch (e: Exception) {
            throw PairingError.Unreachable
        } finally {
            connection.removeSubscriber(subId)
        }
    }

    private fun mapError(message: ServerMessage): PairingError {
        val err = message as? ServerMessage.Error
            ?: return PairingError.Server(-1, "Unexpected reply")
        return when (err.code) {
            401 -> PairingError.InvalidCode
            429 -> PairingError.RateLimited
            else -> PairingError.Server(err.code, err.message)
        }
    }
}
